### PR TITLE
Tentative fix for a race between compute_gc_stats and domain termination

### PR DESCRIPTION
**Pull request by gasche**
This is a tentative fix for a race spotted by @eutro in #12595.

@eutro I would be honored to have your opinion on the proposed fix.

### CODE REVIEW

File: runtime/gc_stats.c
Comment (eutro) [inline runtime/gc_stats.c]: This feels awkward, given that it ignores the parameter and checks the current domain instead (which also means it does a thread local load), contradicting the comment. It should be possible to replicate `caml_domain_is_terminating` for a given domain by looking it up in the `all_domains` table by ID.

**Review by eutro — APPROVED** [review]
This definitely seems like a sensible solution, and looks correct to me at a glance.

File: runtime/gc_stats.c
Comment (gasche) [reply runtime/gc_stats.c]: Good point, thanks. I rebased the PR to follow your suggestion. (I first tried to convince myself that redefining `caml_collect_gc_stats_sample` to take `(void)` and be special-cased on the current domain would be fine; but the rest of the minor GC code is careful to pass the domain state around, so it would feel bad to do this.)

I had trouble naming the `is_terminating` function parametrized on the domain. Clearly it should be called `caml_domain_is_terminating` (and the current one should be called `caml_domain_self_is_terminating` or something), but the `caml_domain_is_foo(void)` pattern is used in several places already. In any case, I named the new one `caml_domain_terminating`, which is not great but acceptable.

**Comment by gasche** [comment]
This needs a decision from a maintainer (who could review or make a decision based on @eutro's review); I wonder if @Engil or @kayceesrk would be available to have a look.

File: runtime/caml/domain.h
Comment (kayceesrk) [inline runtime/caml/domain.h:125]: I wonder whether we can do better with the function names. Given that we now have `caml_domain_terminating` it is not obvious that `caml_domain_is_terminating` refers to the current domain. 

For example, the only place where `caml_domain_is_terminating` is at

https://github.com/ocaml/ocaml/blob/fd0551b7b73e9b65e105a7071642aae155d1ae77/runtime/major_gc.c#L399-L406

It may be best to remove `caml_domain_is_terminating` and only use `caml_domain_terminating` (which would be better named as `caml_domain_is_terminating`.

File: runtime/gc_stats.c
Comment (kayceesrk) [inline runtime/gc_stats.c:107]: Is the "including those happening during domain termination" necessary? 

"each minor collection" ==> "minor collection at domain termination"

File: runtime/gc_stats.c
Comment (kayceesrk) [inline runtime/gc_stats.c:111]: Is the change in name with the addition of the suffix `_stw` indicating that this function should only be called in stop-the-world sections? Is this true for the rest of the runtime? It may be useful to document this convention somewhere.

**Review by kayceesrk — APPROVED** [review]
I believe that the code is doing the right thing. Approved.

File: runtime/caml/domain.h
Comment (gasche) [reply runtime/caml/domain.h:125]: I would be happy with `caml_domain_is_terminating` having an argument, but there are several `caml_domain_is_foo` functions and they all implicitly talk about the current/self domain. (`caml_domain_is_in_multicore`, `caml_domain_is_in_stw`; and also... `caml_domaine_alone ). I didn't want to break the regularity there, so I did the so-so choice of having a minor variant of the function that takes an argument.

I am happy to discuss better naming schemes for these three functions together, but I think this can be done independently of whether the present PR is merged.

For example we could rename those functions into `caml_current_domain_is_foo`, and `caml_domain_is_foo` would take a parameter. Or we could just give up on the `void` version and always pass the domain explicitly. (But I'm not sure they all take a `domain_state`, maybe some rely on `domain_self` instead.)

Comment (gasche) [reply runtime/caml/domain.h:125]: (Passing the domain explicitly would also sometimes be slightly more efficient as it avoids loading the thread-local variable again.)

File: runtime/gc_stats.c
Comment (gasche) [reply runtime/gc_stats.c:107]: Before this change, the function was called either in STW, or at domain termination (but outside the STW sections of `domain_terminate`). It is important for race-freedom that we only do it during STW sections. It is important for correctness that we do *something* at domain termination, otherwise the sampled copy could remain stale forever after termination, so this is why I point those out explicitly.

File: runtime/gc_stats.c
Comment (gasche) [reply runtime/gc_stats.c:111]: Yes, I think all `foo_stw` sections are meant to be called from inside a STW section. (Of course some use the `foo_from_stw`, and some have `stw` somewhere in the name but not at the end). I am not sure where this convention should be documented; I find it relatively obvious, especially if we consistently used the less common but clearer `_from_stw` variant. We could consider a refactoring PR to rename various functions to enforce a single version of this convention.

**Comment by gasche** [comment]
The CI fails on macos with a failure in `parallel/catch_break.ml`:

```
Fatal error: exception Stdlib.Sys.Break
```

This seems unrelated to the current PR, my guess would be that this test could be flaky, but I have not seen it fail before.

**Comment by gasche** [comment]
Current state:

- @kayceesrk' feedback could turn into small refactoring PRs after the fact, let's see where the discussion goes. I plan to merge independently.
- Before merging, I need to look at this CI issue a bit more to ensure that it is not related to the PR.

**Comment by kayceesrk** [comment]
Ok to consider the refactoring in other PRs.

File: runtime/caml/domain.h
Comment (gasche) [reply runtime/caml/domain.h:125]: I proposed something on that space in #12617.

File: runtime/gc_stats.c
Comment (gasche) [reply runtime/gc_stats.c:111]: I propose a PR to define and enforce naming conventions in #12619. Thanks for asking: I was in "minimal change" mode when writing new runtime code, but it actually felt nice to put these conventions on paper and ensure they are consistently followed.

**Comment by gasche** [comment]
I rebased on the current trunk and the `catch_break` failure is gone. My conclusion is that the test indeed is flaky, and in fact there is already a PR by @gadmm (thanks!) to improve it in #12614. I will now go ahead and merge the data race fix. (The followup PRs #12617 and #12619 are up for review and can take their own time; I will rebase them to fix conflicts with the present PR.)

**Pull request merged**
